frontend/test/metabase/scenarios/models/model-actions.cy.spec.js

```diff
@@ -0,0 +1,201 @@
+
+    openActionEditorFor("Delete Order");
+    fillQuery(" AND status = 'pending'");
+    fieldSettings()
+      .findByRole("radiogroup", { name: /Field type/i })
+      .findByLabelText(/Number/i)
+      .should("be.checked");
+    cy.findByRole("button", { name: "Update" }).click();
+
+    cy.findByLabelText("Action list")
+      .findByText(
+        "DELETE FROM orders WHERE id = {{ id }} AND status = 'pending'",
+      )
+      .should("be.visible");
+  });
+
+  it("should allow to make actions public and execute them", () => {
+    cy.intercept("/api/public/action/*/execute", request => {
+      expect(request.body).to.deep.equal({
+        parameters: { [TEST_PARAMETER.id]: -2 },
+      });
+    });
+
+    cy.get("@modelId").then(modelId => {
+      cy.request("POST", "/api/action", {
+        ...SAMPLE_QUERY_ACTION,
+        model_id: modelId,
+      });
+      cy.visit(`/model/${modelId}/detail`);
+      cy.wait("@getModel");
+    });
+
+    cy.findByText("Actions").click();
+    openActionEditorFor(SAMPLE_QUERY_ACTION.name);
+
+    cy.findByRole("dialog").within(() => {
+      cy.findByRole("button", { name: "Action settings" }).click();
+      cy.findByLabelText("Make public").should("not.be.checked").click();
+      cy.findByLabelText("Public action link URL")
+        .invoke("val")
+        .then(url => {
+          cy.wrap(url).as("publicUrl");
+        });
+    });
+
+    cy.get("@publicUrl").then(url => {
+      cy.signOut();
+      cy.visit(url);
+    });
+
+    cy.findByLabelText(TEST_PARAMETER.name).type("-2");
+    cy.findByRole("button", { name: "Submit" }).click();
+    cy.findByText("Thanks for your submission.").should("be.visible");
+    cy.findByRole("form").should("not.exist");
+    cy.findByRole("button", { name: "Submit" }).should("not.exist");
+
+    cy.signInAsAdmin();
+    cy.get("@modelId").then(modelId => {
+      cy.visit(`/model/${modelId}/detail`);
+      cy.wait("@getModel");
+    });
+
+    cy.findByText("Actions").click();
+    openActionEditorFor(SAMPLE_QUERY_ACTION.name);
+
+    cy.findByRole("dialog").within(() => {
+      cy.findByRole("button", { name: "Action settings" }).click();
+      cy.findByLabelText("Make public").should("be.checked").click();
+    });
+    modal().findByRole("button", { name: "Yes" }).click();
+
+    cy.get("@publicUrl").then(url => {
+      cy.visit(url);
+      cy.findByRole("form").should("not.exist");
+      cy.findByRole("button", { name: "Submit" }).should("not.exist");
+      cy.findByText("An error occurred.").should("be.visible");
+    });
+  });
+});
+
+function openActionEditorFor(actionName) {
+  cy.findByRole("listitem", { name: actionName }).within(() => {
+    cy.icon("pencil").click();
+  });
+}
+
+function fillQuery(query) {
+  cy.get(".ace_content").type(query, { parseSpecialCharSequences: false });
+}
+
+function fieldSettings() {
+  cy.findByTestId("action-form-editor").within(() => cy.icon("gear").click());
+  return popover();
+}
```

Comment:
can we extract these elsewhere to be reusable helper functions? maybe make a new set for actions?

Reply:
I'd wait until we actually need them anywhere else (at least in one more file). If we extract them before that, it''d be way easier to end up with dead code

Reply:
Agree, this is what I usually do also when there's not an obvious place where we should put the function or not sure if it should be reuse.